### ISSUE TITLE
2-Factor Auth

### ISSUE THREAD

**Issue by eragon921** [rmcrackan/Libation#49]
When signing in to my account, the CAPTCHA works just fine, but there is no way to complete the 2nd factor authorization with a code from Amazon. Is 2FA support on your list of features to include? I'm loathe to disable 2FA on my Amazon account just to be able to sync with Audible.

**Comment by rmcrackan** [comment]
Agreed. Please do not do that. I'm in the process of fixing this for someone else too. I must have broken it recently. Please try again with [v5.1.4](https://github.com/rmcrackan/Libation/releases/latest). If this doesn't fix it, Please reply here.

**Comment by eragon921** [comment]
Hey sir,

 

So I now get the Two-Step verification prompt asking how I want the code delivered, but pressing Submit does nothing. 

 

S

**Comment by rmcrackan** [comment]
Can you clarify "does nothing"? As in: does that prompt refuse to close? Does it close and the login just attempt stops? Is there an error?

Also, can you share your logs?

**Comment by eragon921** [comment]
It literally does nothing. No pop ups, doesn't close, no error generated. here are the logs.
[Log202107.log](https://github.com/rmcrackan/Libation/files/6787449/Log202107.log)

**Comment by rmcrackan** [comment]
Thanks for the logs. We're definitely getting closer. [v5.1.5](https://github.com/rmcrackan/Libation/releases/tag/v5.1.5) has no bug fixes but I did add a lot of debug logging around that area. Please try 5.1.5 and in your Settings.json file, change the following settings. You may want to make a backup of your file first to more easily reset them later:

Logging LogLevel Default: "Debug"

MinimumLevel: "Debug"

I think/hope your next log will have something interesting to show me.

**Comment by eragon921** [comment]
Here's the new log from 5.1.5. Logging level is set to Debug
[Log202107.log](https://github.com/rmcrackan/Libation/files/6788003/Log202107.log)

**Comment by rmcrackan** [comment]
Thanks for your log file. So sorry -- stupid bug fixed in [latest](https://github.com/rmcrackan/Libation/releases/tag/v5.1.6)

**Comment by rmcrackan** [comment]
Interestingly in your 5.1.5 log, no debug lines were shown (lines with [DBG] instead of [INF]). No idea why. Version 5.1.6 also adds some logging-level info in case this comes up again

**Comment by eragon921** [comment]
Thank you sir! That works now. Book downloading and decrypting is working. 

 

I have one other LOW PRIORITY feature request. If you want me to open another issue, I will. In short, can a setting be added to customize how the output folders and files are named? I’ve been using InAudible and Audible-CLI for so long that I’ve gotten used to being able to set it to Author – Title – Series. This is just a preference thing. Otherwise, I’m quite impressed with what you’ve built! Thank you so much.

**Comment by rmcrackan** [comment]
You're welcome. And thank you for your assistance. And the generous tip. I look forward to buying more audiobooks with it soon :)

Regarding that request: it sounds good on the surface and you're welcome to open a new issue for it. Here's the challenge:

tl;dr: if the audible id is not either in the file or in the folder name, Libation doesn't know what book it is

The long version:

A huge goal is to keep book download/decrypt as simple as possible. I want a 2-click solution: 1 - update my library - 2 - give me all of my new books. The challenge is: how do you know what file goes with which library entry? If I keep a database, it will be out of date the moment a file is moved or renamed. If I am constantly read the file's metadata, it gets *very* CPU hungry especially with large libraries. My compromise is to scan file and folder names only (not to try to read the files) and look for the audible id. Libation assume a match if it either - 1 - finds any folder with that id which contains an audio file (file name unimportant) or - 2 - if it finds any audio file with that id in the file's name. This compromise means that within 1 parent folder, I can have folders categorized and nested any way my little OCD heart desires and Libation can keep the library in sync. I'm open to suggestion. However any new paradigm would have to respect the goals of ease of use, flexible folder hierarchy, and not slowing down the user's computer
